mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -0,0 +1,161 @@
+    let mut tokens = coin.get_erc_tokens_infos();
+
+    // Workaround for performance purposes.
+    //
+    // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
+    // type and mapping the platform coin and the entire token list (which can grow at any time), we map
+    // the platform coin to Erc20TokenInfo so that we can use the token list right away without
+    // additional mapping.
+    tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
+        token_address: coin.my_address,
+        decimals: coin.decimals,
+    });
+
+    let jobs = tokens
+        .into_iter()
+        .map(|(token_ticker, info)| async move {
+            if token_ticker == coin.ticker {
+                let balance_as_u256 = coin
+                    .address_balance(coin.my_address)
+                    .compat()
+                    .await
+                    .map_err(|e| (token_ticker.clone(), e))?;
+                let balance_as_big_decimal = u256_to_big_decimal(balance_as_u256, coin.decimals)
+                    .map_err(|e| (token_ticker.clone(), e.into()))?;
+                Ok((coin.ticker.clone(), balance_as_big_decimal))
+            } else {
+                let balance_as_u256 = coin
+                    .get_token_balance_by_address(info.token_address)
+                    .await
+                    .map_err(|e| (token_ticker.clone(), e))?;
+                let balance_as_big_decimal = u256_to_big_decimal(balance_as_u256, info.decimals)
+                    .map_err(|e| (token_ticker.clone(), e.into()))?;
+                Ok((token_ticker, balance_as_big_decimal))
+            }
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    jobs.collect().await
+}
+
+#[async_trait]
+impl EventBehaviour for EthCoin {
+    const EVENT_NAME: &'static str = "COIN_BALANCE";
+    const ERROR_EVENT_NAME: &'static str = "COIN_BALANCE_ERROR";
+
+    async fn handle(self, interval: f64, tx: oneshot::Sender<EventInitStatus>) {
+        const RECEIVER_DROPPED_MSG: &str = "Receiver is dropped, which should never happen.";
+
+        async fn with_socket(_coin: EthCoin, _ctx: MmArc) { todo!() }
+
+        async fn with_polling(coin: EthCoin, ctx: MmArc, interval: f64) {
+            let mut cache: BTreeMap<String, BigDecimal> = BTreeMap::new();
+
+            loop {
+                let now = Instant::now();
+
+                let mut balance_updates = vec![];
+                for result in get_all_balance_results_concurrently(&coin).await {
+                    match result {
+                        Ok((ticker, balance)) => {
+                            if Some(&balance) == cache.get(&ticker) {
+                                continue;
+                            }
+
+                            balance_updates.push(json!({
+                                "ticker": ticker,
+                                "balance": { "spendable": balance, "unspendable": BigDecimal::default() }
```

Comment:
I see that there is a duplication of this serialization [here](https://github.com/KomodoPlatform/komodo-defi-framework/blob/c92710f96c3ef7ef32a54fcf7008a69c7074dbcd/mm2src/coins/tendermint/tendermint_balance_events.rs#L152-L155)

I suggest to create a struct type and use `serde_json` crate

Reply:
I suggest to define struct that represents the json you are creating, for example
```rust
#[derive(Serialize)]
struct BalanceUpdate {
    ticker: String,
    balance: Balance,
}

#[derive(Serialize)]
struct Balance {
    spendable: BigDecimal,
    unspendable: BigDecimal,
}
```
and instead of using json!, you can create an instance of the `BalanceUpdate` struct and serialize it with `serde_json::to_value()`.

This replaces the manual json construction in `eth_balance_events.rs` and `tendermint_balance_events.rs` with struct-based approach.

Reply:
> This replaces the manual json construction in eth_balance_events.rs and tendermint_balance_events.rs with struct-based approach.

We only need JSON data with just 2 fields. Any operation before obtaining this JSON payload is a waste of time and resources. We don't need a type for this data since we don't use it; we simply publish it as a JSON payload.

Reply:
Ok, got your point. I dont insist, so lets leave it.